src/sql/workbench/services/notebook/browser/notebookServiceImpl.ts

```diff
@@ -437,26 +430,26 @@ export class NotebookService extends Disposable implements INotebookService {
 	//Returns an instantiation of RenderMimeRegistry class
 	getMimeRegistry(): RenderMimeRegistry {
 		if (!this._mimeRegistry) {
-			return new RenderMimeRegistry({
+			this._mimeRegistry = new RenderMimeRegistry({
 				initialFactories: standardRendererFactories
 			});
 		}
 		return this._mimeRegistry;
 	}
 
-	private get providersMemento(): NotebookProvidersMemento {
+	protected get providersMemento(): NotebookProvidersMemento {
```

Comment:
Why are we just making a bunch of things protected? Private vs Protected is a design decision. Testing shouldn't affect that. If you need to change this visiblity to test it, your testing is probably wrong. There's a reason why in c# its considered bad practice to just make everything internal.

Reply:
It was charles' suggestion to make private fields to protected for testing purposes. C# has friend that circumvents this issue.

Reply:
I very much disagree with that, like I said, private vs protected is a design decision not a testing decision. If someone else came and extended this class how are they supposed to know what methods were meant to be used by extended classes vs ones that were just made that way for testing?

Reply:
So let me rephrase what Arvind said with what I was actually saying : 

1. What Anthony said is 100% correct and I agree with it all
1. But I'm also aware that we have not been great about designing components in such a way that they can be easily testable without either a significant refactor, unless we modify private internals. For example some of our stuff is tightly coupled to the UI and so trying to make a test for that ends up being very difficult outside of a smoke test.

Now sure, we could (and have) just access these members via bracket notation. I'm hesitant to do this for two reasons
1. You don't get the compile time errors when a field is changed (so if it's renamed the tests would happily still try to access the old name and only fail then)
1. People modifying the class have no idea that other stuff is using these, which ends up being a pain when you make changes to what should be an implementation detail only to find out you need to go and update a bunch of tests that were modifying that state directly

There's a lot of nuance here - but I think at the end my thoughts are : 

1. We should always be striving to test things without needing to modify/access their internal state. This is just good test design and should be possible for nearly all cases (I think I've had to do this once so far for all the tests I've written)
1. If you do need to - consider instead refactoring the component such that you don't need to modify internal state to achieve your testing goals
1. But in the end we have to balance time spent writing tests with time spent refactoring and the amount we gain from that time spent. And if it ends up being a lot of work for very little gain then I personally see no problem with having a few places expose their internals (or have tests access/modify them).

In this specific case I see way too many private -> protected changes for my liking. I would very much question why we aren't able to test this behavior without making these protected. So @ranasaria if you really believe that these are necessary please go into detail explaining why so people have context and can perhaps suggest alternatives.

Reply:
So these are the two questions that we need to address here: 
1. Should we ever need access to private members in tests? 
2. And if we fathom that it is indeed ok to allow access to private members in testing, then what is the best way to do so in typescript.

Let's address the 1st question. Let me present an example of a common code pattern from our codebase:
1. There exists an EventEmitter on class Foo and its event is available publicly for listeners to subscribe to.
2. The class under test, say 'Bar', subscribes to that public event on Foo and registers an async private method to be invoked when the corresponding emitter fires. The code in Bar or the product does not care when the execution of this async method completes and it raises errors when it runs into any unexpected issues.

So to get this code tested, we can write a test to invoke whatever code will fire the event in question by using the corresponding EventEmitter in Foo class. However, to do verification of the effects of the async private method that is the focus of this test, we need to be able to wait for the completion of this async method. There is no current way available to do this wait. The actual product code does not need to do this wait because there is nothing that needs to be done in the product after that method finishes. By doing access to the private method, one can test this code out in two parts:
1. Write test code to make sure that the private method under test gets invoked when the EventEmitter on Foo fires. 
2. Invoke the async method through direct access so you can await on it and do verification once the method execution has completed.

Let's also address the 2nd question: (this may be redundant if we find a good way to solve the first question without requiring direct access). So if were to allow direct access to private members then what is the best way to do so. Two obvious ways come to mind:
1. Direct member access using string indexers on the corresponding object. 
2. Exposing private members as protected to provide access via a child class.

I have found the 2nd approach to definitely be better because it makes a future developer enhancing the class under test ('Bar, in our above example) be aware of the protected members and potential usages outside of that class and hence allows him/her to do the right thing. Of course, the downside is that the developer still has to worry about the protected members which s/he would not have had to if it had remained private. The first approach actually hides usage of those members completely making life hard for the future developer who is likely to discover breakages of tests later than sooner.

Reply:
@ranasaria Could you please point to an example in the code/tests you wrote that shows where this was needed to be done? It'd help to give direct context to everyone so we don't have to spend time going through your tests to understand what parts you're referring to.

Reply:
Fixed. Removed the Subclass altogether.